### PR TITLE
feat(datastore): support for @auth provider

### PR DESCRIPTION
**Pull request by diegocstn**
*Description of changes:*
This PR introduces changes to capture the value of `provider` in @auth directives as an `AuthRule` metadata.

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [ ] All integration tests pass
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift [hunk 1]
```diff
@@ -23,6 +23,14 @@ public enum ModelOperation {
     case read
 }
 
+/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
+///   by host applications. The behavior of this may change without warning.
+public enum AuthRuleProvider {
+    case apiKey
+    case oidc
+    case custom(name: String)
```
Comment (lawmicha) [inline Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift:31]: is the plan for codegen to generate strings like `.apiKey`, `oidc`, `.custom("iam")`, and `.custom("userPools")` in the Model+Schema file?

Comment (diegocstn) [reply Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift:31]: The plan is to generate enum cases and/or static method calls like `.apiKey`, `.oidc`, `.iam` `.userPools`.
`AuthRuleProvider` contains only generic providers, and `AWSAuthRuleProvider` adds AWS specific providers via static methods but at the call site won't make any difference

Comment (lawmicha) [reply Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift:31]: would the generate code need `import AWSPluginsCore`?

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift [hunk 2]
```diff
@@ -7,6 +7,44 @@
...
+    public func toAWSAuthorizationType() throws -> AWSAuthorizationType {
+        var authType: AWSAuthorizationType
+        switch self {
+        case .apiKey:
+            authType = .apiKey
+        case .oidc:
+            authType = .openIDConnect
+        case .custom(name: let name) where name == AuthRuleProvider.valueForIAM():
+            authType = .awsIAM
+
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift:36]:
```suggestion
```

**Review by lawmicha — APPROVED** [review]

File: Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift (hunk 1 above)
Comment (drochetti) [reply Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift:31]: > would the generated code need `import AWSPluginsCore`?

that's a good question... if we create a dependency between the generated models and `AWSPluginsCore` I wonder if the abstraction is worth it, given I don't think it's realistic to drift away from a non-AWS permission model (a security model that doesn't depend on Cognito would require the entire system to change).

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift [hunk 3]
```diff
@@ -7,6 +7,43 @@
...
+    public static var iam: AuthRuleProvider {
+        .custom(name: valueForIAM())
+    }
+
+    public static var userPools: AuthRuleProvider {
+        .custom(name: valueForCognitoUserPool())
+    }
+
+    public func toAWSAuthorizationType() throws -> AWSAuthorizationType {
```
Comment (drochetti) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift:27]: what is the difference between `AWSAuthorizationType` and `AuthRuleProvider`? Is it the latter constrained to the metadata and the former used in the auth implementation?

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift [hunk 4]
```diff
@@ -7,6 +7,43 @@
...
+        case .oidc:
+            authType = .openIDConnect
+        case .custom(name: let name) where name == AuthRuleProvider.valueForIAM():
+            authType = .awsIAM
+        case .custom(name: let name) where name == AuthRuleProvider.valueForCognitoUserPool():
+            authType = .amazonCognitoUserPools
+        case .custom(name: let name):
+            throw DataStoreError.unknown(
+                "Invalid AWS authorization provider in schema @auth rule: \(name)",
+                "Verify your schema.")
```
Comment (drochetti) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift:41]: thinking out loud... given the need to handle unknown cases and the `import AWSPluginsCore`, I wonder if we should drop the AWS abstraction and just define all the supported cases in the `AuthRuleProvider` enum itself.

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift (hunk 3 above)
Comment (diegocstn) [reply AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift:27]: Yes, that's correct. `AuthRuleProvider` defines values allowed by `@auth` rule specs

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift (hunk 4 above)
Comment (diegocstn) [reply AmplifyPlugins/Core/AWSPluginsCore/Model/Support/AuthRule+Extension.swift:41]: yeah, I agree it's probably easier to drop the AWS abstraction and simplify things a bit

**Pull request merged**